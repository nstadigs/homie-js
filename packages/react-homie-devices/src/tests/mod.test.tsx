import { assertSnapshot } from "@std/testing/snapshot";

// @deno-types="npm:@types/react"
import React, { act, useState } from "react";
import { Device, Node, Property, register } from "../mod.ts";
import { TestMqttAdapter } from "./testutils.ts";

Deno.test("Initial render", async (t) => {
  // deno-lint-ignore require-await
  const { cleanUp, mqtt } = await act(async () => {
    return testRender(<Controller />);
  });

  await assertSnapshot(t, mqtt.events);
  await act(cleanUp);
});

Deno.test("Set new value and target", async (t) => {
  const { cleanUp, mqtt } = await act(() => {
    return testRender(<Controller />);
  });

  mqtt.events = [];

  await act(() => {
    mqtt.publish(
      "homie/5/root-device/root-device-node/property-2/set",
      "41",
      0,
      false,
    );
  });

  await assertSnapshot(t, mqtt.events);
  await act(cleanUp);
});

Deno.test("Set updates root device", async (t) => {
  function Controller() {
    const [rootDeviceName, setRootDeviceName] = useState("");

    return (
      <Device id="root-device" name={`${rootDeviceName}`}>
        <Node id="root-device-node">
          <Property
            id="device-name-property"
            datatype="string"
            onSet={(name) => {
              setRootDeviceName(name);
            }}
          />
        </Node>
        <Device id="child-device" />
      </Device>
    );
  }

  const { cleanUp, mqtt } = await act(() => {
    return testRender(<Controller />);
  });

  mqtt.events = [];

  await act(() => {
    mqtt.publish(
      "homie/5/root-device/root-device-node/device-name-property/set",
      "New name",
      0,
      false,
    );
  });

  await assertSnapshot(t, mqtt.events);
  await act(cleanUp);
});

Deno.test("Set updates property", async (t) => {
  function Controller() {
    const [propertyName, setPropertyName] = useState("old name");

    return (
      <Device id="root-device" name={`${propertyName}`}>
        <Node id="root-device-node">
          <Property
            id="property-name"
            name={propertyName}
            datatype="string"
            onSet={(name) => {
              setPropertyName(name);
            }}
          />
        </Node>
        <Device id="child-device" />
      </Device>
    );
  }

  const { cleanUp, mqtt } = await act(() => {
    return testRender(<Controller />);
  });

  mqtt.events = [];

  await act(() => {
    mqtt.publish(
      "homie/5/root-device/root-device-node/property-name/set",
      "New name",
      0,
      false,
    );
  });

  await assertSnapshot(t, mqtt.events);
  await act(cleanUp);
});

function Controller() {
  const [someValue, setSomeValue] = useState(0);

  return (
    // TODO: Name
    <Device id="root-device" name="0">
      <Node id="root-device-node">
        <Property
          id="property-2"
          name="Property 3"
          datatype="integer"
          onSet={(addValue) => {
            setSomeValue((value) => value + addValue);
          }}
          value={someValue}
          target={someValue}
        />
      </Node>
      <Device id="child-device" />
    </Device>
  );
}

function testRender(whatToRender: React.ReactNode) {
  // TODO: This only returns a single mqtt adapter. We should make
  // one per root devices somehow.
  const mqtt = new TestMqttAdapter();
  const cleanUp = register(whatToRender, () => mqtt);
  return { mqtt, cleanUp };
}
