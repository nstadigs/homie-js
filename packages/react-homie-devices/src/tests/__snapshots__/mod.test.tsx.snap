export const snapshot = {};

snapshot[`Initial render 1`] = `
[
  {
    payload: "0",
    qos: 2,
    retained: false,
    topic: "homie/5/root-device/root-device-node/property-2",
  },
  {
    payload: "0",
    qos: 2,
    retained: false,
    topic: "homie/5/root-device/root-device-node/property-2/\$target",
  },
  {
    payload: "init",
    qos: 2,
    retained: true,
    topic: "\$state",
  },
  {
    payload: '{"homie":"5.0","version":1145439745,"name":"child-device","root":"root-device","parent":"root-device","children":[],"nodes":{}}',
    qos: 2,
    retained: true,
    topic: "homie/5/child-device/\$description",
  },
  {
    payload: "ready",
    qos: 2,
    retained: true,
    topic: "\$state",
  },
  {
    payload: "init",
    qos: 2,
    retained: true,
    topic: "\$state",
  },
  {
    payload: '{"homie":"5.0","version":1623117270,"name":"0","children":["child-device"],"nodes":{"root-device-node":{"name":"root-device-node","properties":{"property-2":{"name":"Property 3","datatype":"integer","settable":true}}}}}',
    qos: 2,
    retained: true,
    topic: "homie/5/root-device/\$description",
  },
  {
    payload: "ready",
    qos: 2,
    retained: true,
    topic: "\$state",
  },
]
`;

snapshot[`Set new value and target 1`] = `
[
  {
    payload: "41",
    qos: 0,
    retained: false,
    topic: "homie/5/root-device/root-device-node/property-2/set",
  },
  {
    payload: "41",
    qos: 2,
    retained: false,
    topic: "homie/5/root-device/root-device-node/property-2",
  },
  {
    payload: "41",
    qos: 2,
    retained: false,
    topic: "homie/5/root-device/root-device-node/property-2/\$target",
  },
]
`;

snapshot[`Set updates root device 1`] = `
[
  {
    payload: "New name",
    qos: 0,
    retained: false,
    topic: "homie/5/root-device/root-device-node/device-name-property/set",
  },
  {
    payload: "init",
    qos: 2,
    retained: true,
    topic: "homie/5/root-device/\$state",
  },
  {
    payload: '{"homie":"5.0","version":172139472,"name":"New name","children":["child-device"],"nodes":{"root-device-node":{"name":"root-device-node","properties":{"device-name-property":{"name":"device-name-property","datatype":"string","settable":true}}}}}',
    qos: 2,
    retained: true,
    topic: "homie/5/root-device/\$description",
  },
  {
    payload: "ready",
    qos: 2,
    retained: true,
    topic: "homie/5/root-device/\$state",
  },
]
`;

snapshot[`Set updates property 1`] = `
[
  {
    payload: "New name",
    qos: 0,
    retained: false,
    topic: "homie/5/root-device/root-device-node/property-name/set",
  },
  {
    payload: "init",
    qos: 2,
    retained: true,
    topic: "homie/5/root-device/\$state",
  },
  {
    payload: '{"homie":"5.0","version":4138580612,"name":"New name","children":["child-device"],"nodes":{"root-device-node":{"name":"root-device-node","properties":{"property-name":{"name":"New name","datatype":"string","settable":true}}}}}',
    qos: 2,
    retained: true,
    topic: "homie/5/root-device/\$description",
  },
  {
    payload: "ready",
    qos: 2,
    retained: true,
    topic: "homie/5/root-device/\$state",
  },
]
`;
